tests/handlers/test_sync.py

```diff
@@ -186,6 +187,139 @@ def test_unknown_room_version(self):
         self.assertNotIn(invite_room, [r.room_id for r in result.invited])
         self.assertNotIn(knock_room, [r.room_id for r in result.knocked])
 
+    def test_ban_wins_race_with_join(self):
+        """Check rooms appear under the "leave" section if they lose a race to a ban.
+
+        A complicated edge case. Best to quote RvdH from
+        https://github.com/matrix-org/synapse/pull/11532#discussion_r769104461:
```

Comment:
I'd rather ensure we have the full contents described here, instead of linking off to a GitHub comment. Below seems verbatim, so I'm not sure the link offers much more.